File: prelude/toolchains/android/test/com/facebook/buck/jvm/java/ClassUsageMergerTest.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under both the MIT license found in the
 * LICENSE-MIT file in the root directory of this source tree and the Apache
 * License, Version 2.0 found in the LICENSE-APACHE file in the root directory
 * of this source tree.
 */

package com.facebook.buck.jvm.java;

import static org.junit.Assert.assertEquals;

import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.Map;
import org.junit.Test;
import toolchains.android.src.com.facebook.buck.jvm.kotlin.compilerplugins.usedclasses.ClassUsageMerger;

public class ClassUsageMergerTest {

  @Test
  public void testEmptyPreviousClassUsage() {
    Map<Path, Map<Path, Integer>> prevClassUsageMap = new HashMap<>();
    Map<Path, Map<Path, Integer>> currentClassUsageMap = new HashMap<>();
    currentClassUsageMap.put(Paths.get("A.jar"), new HashMap<>());

    Map<Path, Map<Path, Integer>> mergedMap =
        ClassUsageMerger.mergeClassUsageMaps(prevClassUsageMap, currentClassUsageMap);

    assertEquals(currentClassUsageMap, mergedMap);
  }

  @Test
  public void testEmptyCurrentClassUsage() {
    Map<Path, Map<Path, Integer>> prevClassUsageMap = new HashMap<>();
    prevClassUsageMap.put(Paths.get("A.jar"), new HashMap<>());
    Map<Path, Map<Path, Integer>> currentClassUsageMap = new HashMap<>();

    Map<Path, Map<Path, Integer>> mergedMap =
        ClassUsageMerger.mergeClassUsageMaps(prevClassUsageMap, currentClassUsageMap);

    assertEquals(prevClassUsageMap, mergedMap);
  }

  @Test
  public void testMergeClassUsageWhenOuterAddition() {
    Map<Path, Map<Path, Integer>> prevClassUsageMap = getClassUsageMapBase();
    Map<Path, Map<Path, Integer>> currentClassUsageMap = getClassUsageMapBase();
    Map<Path, Integer> innerMap = new HashMap<>();
    innerMap.put(Paths.get("C.class"), 1);
    currentClassUsageMap.put(Paths.get("new-class-abi.jar"), innerMap);

    Map<Path, Map<Path, Integer>> mergedMap =
        ClassUsageMerger.mergeClassUsageMaps(prevClassUsageMap, currentClassUsageMap);

    assertEquals(currentClassUsageMap, mergedMap);
  }

  @Test
  public void testMergeClassUsageWhenInnerAddition() {
    Map<Path, Map<Path, Integer>> prevClassUsageMap = getClassUsageMapBase();
    Map<Path, Map<Path, Integer>> currentClassUsageMap = getClassUsageMapBase();
    currentClassUsageMap.get(Paths.get("class-abi.jar")).put(Paths.get("C.class"), 1);

    Map<Path, Map<Path, Integer>> mergedMap =
        ClassUsageMerger.mergeClassUsageMaps(prevClassUsageMap, currentClassUsageMap);

    assertEquals(currentClassUsageMap, mergedMap);
  }

  @Test
  public void testMergeClassUsageWhenOuterDeletion() {
    Map<Path, Map<Path, Integer>> prevClassUsageMap = getClassUsageMapBase();
    Map<Path, Map<Path, Integer>> currentClassUsageMap = getClassUsageMapBase();
    currentClassUsageMap.remove(Paths.get("class-abi.jar"));

    Map<Path, Map<Path, Integer>> mergedMap =
        ClassUsageMerger.mergeClassUsageMaps(prevClassUsageMap, currentClassUsageMap);

    assertEquals(prevClassUsageMap, mergedMap);
  }

  @Test
  public void testMergeClassUsageWhenInnerDeletion() {
    Map<Path, Map<Path, Integer>> prevClassUsageMap = getClassUsageMapBase();
    Map<Path, Map<Path, Integer>> currentClassUsageMap = getClassUsageMapBase();
    currentClassUsageMap
        .get(Paths.get("symlink_kotlin-stdlib-2.0.0.jar-class-abi.jar"))
        .remove(Paths.get("kotlin/Unit.class"));

    Map<Path, Map<Path, Integer>> mergedMap =
        ClassUsageMerger.mergeClassUsageMaps(prevClassUsageMap, currentClassUsageMap);

    assertEquals(prevClassUsageMap, mergedMap);
  }

  @Test
  public void testMergeClassUsageWhenNoChange() {
    Map<Path, Map<Path, Integer>> prevClassUsageMap = getClassUsageMapBase();
    Map<Path, Map<Path, Integer>> currentClassUsageMap = getClassUsageMapBase();

    Map<Path, Map<Path, Integer>> mergedMap =
        ClassUsageMerger.mergeClassUsageMaps(prevClassUsageMap, currentClassUsageMap);

    assertEquals(prevClassUsageMap, mergedMap);
    assertEquals(currentClassUsageMap, mergedMap);
  }

  private Map<Path, Map<Path, Integer>> getClassUsageMapBase() {
    Map<Path, Map<Path, Integer>> outerMap = new HashMap<>();
    Map<Path, Integer> innerMap1 = new HashMap<>();
    innerMap1.put(Paths.get("kotlin/Unit.class"), 1);
    innerMap1.put(Paths.get("kotlin/io/ConsoleKt.class"), 1);
    innerMap1.put(Paths.get("kotlin/internal/InlineOnly.class"), 1);
    innerMap1.put(Paths.get("kotlin/annotation/Target.class"), 1);
    innerMap1.put(Paths.get("kotlin/annotation/MustBeDocumented.class"), 1);
    innerMap1.put(Paths.get("kotlin/annotation/Retention.class"), 1);
    outerMap.put(Paths.get("symlink_kotlin-stdlib-2.0.0.jar-class-abi.jar"), innerMap1);

    Map<Path, Integer> innerMap2 = new HashMap<>();
    innerMap2.put(Paths.get("java/lang/Object.class"), 1);
    innerMap2.put(Paths.get("java/lang/annotation/Annotation.class"), 1);
    outerMap.put(Paths.get("android.jar"), innerMap2);

    Map<Path, Integer> innerMap3 = new HashMap<>();
    innerMap3.put(Paths.get("D.class"), 1);
    outerMap.put(Paths.get("class-abi.jar"), innerMap3);

    return new HashMap<>(outerMap);
  }
}

File: prelude/toolchains/android/test/com/facebook/buck/jvm/kotlin/compilerplugins/usedclasses/ClassUsageMergerTest.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is dual-licensed under either the MIT license found in the
 * LICENSE-MIT file in the root directory of this source tree or the Apache
 * License, Version 2.0 found in the LICENSE-APACHE file in the root directory
 * of this source tree. You may select, at your option, one of the
 * above-listed licenses.
 */

package com.facebook.buck.jvm.kotlin.compilerplugins.usedclasses

import java.nio.file.Path
import java.nio.file.Paths
import java.util.HashMap
import org.junit.Assert.assertEquals
import org.junit.Test

private typealias ClassUsageMap = Map<Path, Set<Path>>

internal class ClassUsageMergerTest {

  @Test
  fun `when previous class usage is empty then current class usage is returned`() {
    val prevClassUsageMap: ClassUsageMap = HashMap()
    val currentClassUsageMap: ClassUsageMap = mapOf(Paths.get("A.jar") to HashSet())

    val mergedMap: ClassUsageMap = mergeClassUsageMaps(prevClassUsageMap, currentClassUsageMap)

    assertEquals(currentClassUsageMap, mergedMap)
  }

  @Test
  fun `when current class usage is empty then previous class usage is returned`() {
    val prevClassUsageMap: ClassUsageMap = mapOf(Paths.get("A.jar") to HashSet())
    val currentClassUsageMap: ClassUsageMap = HashMap()

    val mergedMap: ClassUsageMap = mergeClassUsageMaps(prevClassUsageMap, currentClassUsageMap)

    assertEquals(prevClassUsageMap, mergedMap)
  }

  @Test
  fun `when outer addition is made then new class usage is included in merged map`() {
    val prevClassUsageMap: ClassUsageMap = classUsageMapBase

    val innerMap: MutableSet<Path> = mutableSetOf(Paths.get("C.class"))
    val currentClassUsageMap =
        classUsageMapBase.apply { toMutableMap()[Paths.get("new-class-abi.jar")] = innerMap }

    val mergedMap: ClassUsageMap = mergeClassUsageMaps(prevClassUsageMap, currentClassUsageMap)

    assertEquals(currentClassUsageMap, mergedMap)
  }

  @Test
  fun `when inner addition is made then new class is included in merged inner map`() {
    val prevClassUsageMap: ClassUsageMap = classUsageMapBase
    val currentClassUsageMap: ClassUsageMap =
        classUsageMapBase.apply {
          toMutableMap()[Paths.get("class-abi.jar")] = setOf(Paths.get("C.class"))
        }

    val mergedMap: ClassUsageMap = mergeClassUsageMaps(prevClassUsageMap, currentClassUsageMap)

    assertEquals(currentClassUsageMap, mergedMap)
  }

  @Test
  fun `when outer deletion is made then deleted class usage is still included in merged map`() {
    val prevClassUsageMap: ClassUsageMap = classUsageMapBase
    val currentClassUsageMap: ClassUsageMap =
        classUsageMapBase.apply { toMutableMap().remove(Paths.get("class-abi.jar")) }

    val mergedMap: ClassUsageMap = mergeClassUsageMaps(prevClassUsageMap, currentClassUsageMap)

    assertEquals(prevClassUsageMap, mergedMap)
  }

  @Test
  fun `when inner deletion is made then deleted class is still included in merged inner map`() {
    val prevClassUsageMap: ClassUsageMap = classUsageMapBase
    val currentClassUsageMap: ClassUsageMap =
        classUsageMapBase.toMutableMap().apply {
          getValue(Paths.get("symlink_kotlin-stdlib-2.0.0.jar-class-abi.jar"))
              .toMutableSet()
              .remove(Paths.get("kotlin/Unit.class"))
        }

    val mergedMap: ClassUsageMap = mergeClassUsageMaps(prevClassUsageMap, currentClassUsageMap)

    assertEquals(prevClassUsageMap, mergedMap)
  }

  @Test
  fun `when no changes are made then merged map is equal to both previous and current class usage maps`() {
    val prevClassUsageMap: ClassUsageMap = classUsageMapBase
    val currentClassUsageMap: ClassUsageMap = classUsageMapBase

    val mergedMap: ClassUsageMap = mergeClassUsageMaps(prevClassUsageMap, currentClassUsageMap)

    assertEquals(prevClassUsageMap, mergedMap)
    assertEquals(currentClassUsageMap, mergedMap)
  }

  private val classUsageMapBase: ClassUsageMap
    get() {
      val outerMap: MutableMap<Path, MutableSet<Path>> = HashMap()
      val innerSet: MutableSet<Path> =
          mutableSetOf(
              Paths.get("kotlin/Unit.class"),
              Paths.get("kotlin/io/ConsoleKt.class"),
              Paths.get("kotlin/internal/InlineOnly.class"),
              Paths.get("kotlin/annotation/Target.class"),
              Paths.get("kotlin/annotation/MustBeDocumented.class"),
              Paths.get("kotlin/annotation/Retention.class"))

      outerMap[Paths.get("symlink_kotlin-stdlib-2.0.0.jar-class-abi.jar")] = innerSet

      val innerSet2: MutableSet<Path> =
          mutableSetOf(
              Paths.get("java/lang/Object.class"),
              Paths.get("java/lang/annotation/Annotation.class"))
      outerMap[Paths.get("android.jar")] = innerSet2

      val innerSet3: MutableSet<Path> = mutableSetOf(Paths.get("D.class"))
      outerMap[Paths.get("class-abi.jar")] = innerSet3

      return outerMap
    }
}
